### ISSUE TITLE
systemd plugin reproducibly gets stuck at the same record. fails silently

### ISSUE THREAD

**Issue by joshgc** [fluent/fluent-bit#6596]
## Bug Report

**Describe the bug**
systemd plugin reproducibly gets stuck at the same record. fails to die or produce any useful warning that it stopped forwarding logs

**To Reproduce**
I'm not sure. I'm hoping you all know how to extract a likely bad record. fluent-bits own output to stdout and stderr has no interesting logs except this last one:

```
[2022/12/20 23:55:59] [ info] [sp] stream processor started
[2022/12/20 23:59:22] [error] [filter:lua:lua.3] error code 2: /fluent-bit/etc/extras.lua:10: attempt to concatenate field '_SOURCE_REALTIME_TIMESTAMP' (a nil value)
```
However in the past any lua errors did not stop the pipeline. Nor does fluent-bit think anything is wrong with the pipelines AFAICT.

- Steps to reproduce the problem:
I stop fluent-bit, delete the database as specified in Systemd:DB. I restart Fluent-Bit. Its not tailing from logs so it digests 10ks of records from journald. I see records grow for several minutes while polling 0.0.0.0:2020/api/v1/metrics . Eventually the # of systemd records stops growing. Permanently.

I manually probe the fluent-bit DB file and see that a) its still being updated and b) the cursor is still.
```
$ sqlite3 /var/lib/fluent-bit-storage/jou* "SELECT * FROM in_systemd_cursor;"
s=d39672aedd2f49309f2fc5c21bbda06d;i=97966dd;b=103a429c4a7f4a34aa15a72277ad4aba;m=43b7f73736c;t=5efd059d23595;x=70ad2198472e4aaa|1671581511

sqlite3 /var/lib/fluent-bit-storage/jou* "SELECT * FROM in_systemd_cursor;"; date +%s
s=d39672aedd2f49309f2fc5c21bbda06d;i=97966dd;b=103a429c4a7f4a34aa15a72277ad4aba;m=43b7f73736c;t=5efd059d23595;x=70ad2198472e4aaa|1671581528
1671581529
```

**Expected behavior**
1) Better reporting of the issue. I think this should maybe include killing the process since its no longer fulfilling its lifes goal of forwarding logs. 
2) Of course it'd be great if it didnt' get stuck at all. If we can correctly digest the record then yay. Otherwise lets skip it and increment some counter of skipped records?

**Your Environment**

* Version used: 1.9.7 and 2.0.6 both get stuck on debian and debian-backports image
* Configuration:
```
...
    # File system buffering.
    storage.path              /var/lib/fluent-bit-storage/
    storage.sync              normal
    storage.checksum          on
    storage.backlog.mem_limit 5M
    storage.metrics           on
...

[INPUT]
    Name            systemd
    Tag             systemd.*
    Path            /var/log/journal
    Read_From_Tail  Off
    DB              /var/lib/fluent-bit-storage/journal_systemd_db
```
* Environment name and version (e.g. Kubernetes? What version?): Running iinside podman
* Server type and version: 
* Operating System and version: host was a custom debian with this version of libsystemd

```
apt-cache showpkg libsystemd0
...
Dependencies: 
252.1-1+gl0 - libc6 (2 2.34) libcap2 (2 1:2.10) libgcrypt20 (2 1.10.0) liblz4-1 (2 0.0~r122) liblzma5 (2 5.1.1alpha+20120614) libzstd1 (2 1.5.2) libsystemd0:i386 (35 252.1-1+gl0) libsystemd0:i386 (38 252.1-1+gl0) 
Provides: 
252.1-1+gl0 - 
``` 
* Filters and plugins: systemd

**Additional context**

I'm scarily broke. Our fleet lost systemd forwarding for a week and no one noticed. Woof. 

Happy Holidays!

**Comment by joshgc** [comment]
**tl;dr**
The cursor gets stuck right at a journal rotation. journald complains the headers are out of date.

I ran
```
journalctl --cursor="$(sqlite3 /var/lib/fluent-bit-storage/jou* "SELECT cursor FROM in_systemd_cursor;")" -o json --until "2022-12-14 21:25:00"
``` 
to get a sample of the journald records. I attached a small subsample for you on this issue. The sweet highlight is

```
SYSLOG_IDENTIFIER,MESSAGE
"systemd-journald","Journal stopped"
"systemd-journald","Received SIGTERM from PID 1 (systemd)."
"systemd","Stopping Journal Service..."
"systemd","systemd-journald.service: Deactivated successfully."
"systemd","Stopped Journal Service."
"systemd","systemd-journald.service: Consumed 1h 2min 1.357s CPU time."
"systemd","Starting Journal Service..."
"systemd-journald","/var/log/journal/f35aa5617cee475c9039a52463426bdd/system.journal uses an outdated header, suggesting rotation."
"systemd-journald","/var/log/journal/f35aa5617cee475c9039a52463426bdd/system.journal: Journal header limits reached or header out-of-date, rotating."
"systemd-journald","Journal started"
"systemd-journald","System Journal (/var/log/journal/f35aa5617cee475c9039a52463426bdd) is 2.7G, max 4.0G, 1.2G free."
```

**Comment by joshgc** [comment]
[dumpledump.txt](https://github.com/fluent/fluent-bit/files/10280463/dumpledump.txt)

Which was produced via
```
cat journald_records.json | head -n50 | jq -r '[.SYSLOG_IDENTIFIER,.MESSAGE] | @csv'
# I will uselessly use cat until I die.
```

**Comment by chitoku-k** [comment]
Hi, I also have the same issue. FYI, I can reproduce this on systemd v252 but not on v251. Some internal changes they made might have broken this input plugin; however, I currently have no idea as to which commit is the source of this stuck.

**Comment by joshgc** [comment]
Bumping post holidays because I'm still receiving outage notices and am not sure how to fix the issue

**Comment by joshgc** [comment]
I do think we need better reporting from the systemd plugin here BUT

I'm no longer blocked. I was able to roll our container forward to debian bookworm which has version 252 of libsystemd0 and that has started forwarding logs again.

**Comment by chitoku-k** [comment]
It seems that a new behavior in systemd v252 described below caused fluent-bit unable to fetch records.

> Journal files gained a new compatibility flag 'HEADER_INCOMPATIBLE_COMPACT'. Files with this flag implement changes to the storage format that allow reducing size on disk. As with other compatibility flags, older journalctl versions will not be able to read journal files using this new format. The environment variable 'SYSTEMD_JOURNAL_COMPACT=0' can be passed to systemd-journald to disable this functionality.

https://github.com/systemd/systemd/blob/v252/NEWS#L412-L418

I tried the following and no longer find any issues with systemd v252:
```bash
cat <<EOF > /etc/systemd/system/systemd-journald.service.d/override.conf
[Service]
Environment=SYSTEMD_JOURNAL_COMPACT=0
EOF

systemctl daemon-reload
systemctl restart systemd-journald
```

**Comment by chitoku-k** [comment]
It is now fixed in [v2.0.9](https://github.com/fluent/fluent-bit/tree/v2.0.9) as it includes [libsystemd0](https://packages.debian.org/en/bullseye-backports/libsystemd0) 252.4-1~bpo11+1.
